index.d.ts

```diff
@@ -1,288 +1,620 @@
-declare module 'react-table' {
-  import { ReactNode, useState } from 'react'
-
-  type StringKey<D> = Extract<keyof D, string>
-  type IdType<D> = StringKey<D> | string
+// TypeScript Version: 3.0
 
-  type SortingRule<D> = {
-    id: IdType<D>
-    desc: boolean
+declare module 'react-table' {
+  import { ReactNode, useState, ComponentType, MouseEvent, Props } from 'react'
+
+  export interface TableOptions<D extends object> extends UseTableOptions<D> {
+    //,
+    // UseSortByOptions<D>,
+    // UseFiltersOptions<D>,
+    // UseGroupByOptions<D>,
+    // UseExpandedOptions<D>,
+    // UsePaginationOptions<D>,
+    // UseRowSelectOptions<D>,
+    // UseRowStateOptions<D> {
+    // columns: [] // FIXME
+    // state: [] //FIXME
   }
 
-  export type SortingRules<D> = SortingRule<D>[]
+  export interface TableInstance<D extends object = {}>
+    extends TableOptions<D>,
+      UseTableInstanceProps<D> {}
+
+  /**
+   * The empty definition of TableState provides a definition for the state, that can then be extended in the users code.
+   *
+   * @example
+   *  export interface TableState<D extends object = {}}>
+   *    extends UsePaginationState,
+   *      UseGroupByState,
+   *      UseSortbyState<D>,
+   *      UseFiltersState<D> {}
+   */
+  export interface TableState<D extends object = {}> {} // eslint-disable-line @typescript-eslint/no-empty-interface
+
+  export interface Hooks<D extends object = {}> extends UseTableHooks<D> {} //&
+  // UseSortByHooks<D> &
+  // UseGroupByHooks<D> &
+  // UseExpandedHooks<D> &
+  // UseRowSelectHooks<D>
+
+  export interface Cell<D extends object = {}> extends UseTableCellProps<D> {}
+
+  export interface Column<D extends object = {}>
+    extends UseTableColumnOptions<D> {} //& UseGroupByColumnOptions<D> & UseFiltersColumnOptions<D>
+
+  export interface ColumnInstance<D extends object = {}>
+    extends Column<D>,
+      UseTableColumnProps<D> {} //& UseGroupByColumnProps<D> & UseFiltersColumnProps<D>
+
+  export interface HeaderGroup<D extends object = {}>
+    extends UseTableHeaderGroupProps<D>,
+      ColumnInstance<D> {}
+
+  export interface Row<D extends object = {}> extends UseTableRowProps<D> {}
+
+  /* #region useTable */
+  export function useTable<D extends object = {}>(
+    options: TableOptions<D>,
+    ...plugins: PluginHook<D>[]
+  ): TableInstance<D>
 
-  export type SortByFn = (a: any, b: any, desc: boolean) => 0 | 1 | -1
+  export type UseTableOptions<D extends object> = Partial<{
+    // columns: Column<D>[] // FIXME
```

Comment:
Pretty sure that these two need to be uncommented

Reply:
Yeah, I'm going to be getting columns and state fixed while working on the kitchen sink example.